### PR TITLE
chore(sanitize): new implementation of the html sanitized parser

### PR DESCRIPTION
**Pull request by mhevery**
This implementation is based on inert document

[Review on Reviewable](https://reviewable.io/reviews/angular/angular.js/11790)

### CODE REVIEW

File: test/ngSanitize/sanitizeSpec.js [hunk 1]
```diff
@@ -1,6 +1,6 @@
 'use strict';
 
-describe('HTML', function() {
+ddescribe('HTML', function() {
```
Comment (gkalpak) [inline test/ngSanitize/sanitizeSpec.js]: _ddescribe_

File: src/ngSanitize/sanitize.js [hunk 2]
```diff
@@ -262,11 +236,34 @@ function makeMap(str, lowercaseKeys) {
   return obj;
 }
 
+var baseNode;
+(function(window) {
+  var doc;
+  if (window.DOMDocument) {
+    doc = new window.DOMDocument();
```
Comment (lgalfaso) [inline src/ngSanitize/sanitize.js]: Is a `DOMDocument` *inert*? I was not able to find anything that claims this

File: test/ngSanitize/sanitizeSpec.js [hunk 3]
```diff
@@ -88,15 +67,15 @@ describe('HTML', function() {
     });
 
     it('should throw badparse if text content contains "<" followed by "/" without matching ">"', function() {
```
Comment (lgalfaso) [inline test/ngSanitize/sanitizeSpec.js]: given that this is not throwing, the test should be renamed

File: test/ngSanitize/sanitizeSpec.js [hunk 4]
```diff
@@ -88,15 +67,15 @@ describe('HTML', function() {
...
-      expect(function() {
-        htmlParser('foo </ bar', handler);
-      }).toThrowMinErr('$sanitize', 'badparse', 'The sanitizer was unable to parse the following block of html: </ bar');
+      htmlParser('foo </ bar', handler);
+      expect(start).toEqual(undefined);
+      expect(text).toEqual('foo ');
     });
 
     it('should throw badparse if text content contains "<" followed by an ASCII letter without matching ">"', function() {
```
Comment (lgalfaso) [inline test/ngSanitize/sanitizeSpec.js]: same

File: src/ngSanitize/sanitize.js [hunk 5]
```diff
@@ -262,11 +236,34 @@ function makeMap(str, lowercaseKeys) {
...
+  var bodys = docElement.getElementsByTagName('body');
+  if (bodys.length === 1) {
+    baseNode = bodys[0];
+  } else {
+    var html = doc.createElement('html');
+    var body = baseNode = doc.createElement('body');
+    html.appendChild(body);
+    doc.appendChild(html);
+  }+
+})(window);
```
Comment (gkalpak) [inline src/ngSanitize/sanitize.js]: What would be the drawback of using `$window` instead (after moving this inside `$get()`) ?

File: src/ngSanitize/sanitize.js [hunk 6]
```diff
@@ -262,11 +236,34 @@ function makeMap(str, lowercaseKeys) {
...
+  var docElement = doc.documentElement || doc.getDocumentElement();
+  var bodys = docElement.getElementsByTagName('body');
+  if (bodys.length === 1) {
+    baseNode = bodys[0];
+  } else {
+    var html = doc.createElement('html');
+    var body = baseNode = doc.createElement('body');
+    html.appendChild(body);
+    doc.appendChild(html);
+  }+
```
Comment (campersau) [inline src/ngSanitize/sanitize.js]: Is the `+` behind `}` intended?

File: src/ngSanitize/sanitize.js [hunk 7]
```diff
@@ -262,11 +236,34 @@ function makeMap(str, lowercaseKeys) {
...
+    doc = new window.DOMDocument();
+  } else if (window.document && window.document.implementation) {
+    doc = window.document.implementation.createHTMLDocument("inert");
+  } else if (window.ActiveXObject) {
+    doc = new ActiveXObject("Msxml.DOMDocument");
+  } else {
+    throw "DOMDocument not supported!";
+  }
+  var docElement = doc.documentElement || doc.getDocumentElement();
+  var bodys = docElement.getElementsByTagName('body');
```
Comment (IgorMinar) [inline src/ngSanitize/sanitize.js]: bodys -> bodyElements

File: src/ngSanitize/sanitize.js (hunk 6 above)
Comment (IgorMinar) [inline src/ngSanitize/sanitize.js]: +1

File: src/ngSanitize/sanitize.js [hunk 8]
```diff
@@ -262,11 +236,34 @@ function makeMap(str, lowercaseKeys) {
...
+  if (bodys.length === 1) {
+    baseNode = bodys[0];
+  } else {
+    var html = doc.createElement('html');
+    var body = baseNode = doc.createElement('body');
+    html.appendChild(body);
+    doc.appendChild(html);
```
Comment (IgorMinar) [inline src/ngSanitize/sanitize.js]: so we just append it as is since we assume that it's inert?

File: src/ngSanitize/sanitize.js (hunk 5 above)
Comment (IgorMinar) [inline src/ngSanitize/sanitize.js]: there is no drawback, but I don't think there is any benefit either. we need to test this against the real window.

File: src/ngSanitize/sanitize.js (hunk 2 above)
Comment (mhevery) [inline src/ngSanitize/sanitize.js]: That is my understanding, and also I was able to verify this imperatively. If you try to create ``  in regular document you can see either the alert or that it was canceled in the network tab. Creating the same thing on the new document does nothing.

File: src/ngSanitize/sanitize.js (hunk 8 above)
Comment (mhevery) [inline src/ngSanitize/sanitize.js]: So some platforms create `html` > `body` but others do not. This is just to normalize it.

File: src/ngSanitize/sanitize.js (hunk 6 above)
Comment (mhevery) [inline src/ngSanitize/sanitize.js]: Nope a mess up, cleaned up.

File: src/ngSanitize/sanitize.js (hunk 5 above)
Comment (mhevery) [inline src/ngSanitize/sanitize.js]: I thought that `$window` is only in the main module, but maybe I am mistaken.

File: test/ngSanitize/sanitizeSpec.js (hunk 1 above)
Comment (mhevery) [inline test/ngSanitize/sanitizeSpec.js]: removed

**Comment by mhevery** [comment]
@IgorMinar I don't know how to fix the build. Can you have a look?

File: src/ngSanitize/sanitize.js (hunk 5 above)
Comment (gkalpak) [inline src/ngSanitize/sanitize.js]: I meant get it through DI. But as @igor mentioned there isn't any real benefit either, so...

File: src/ngSanitize/sanitize.js (hunk 2 above)
Comment (lgalfaso) [inline src/ngSanitize/sanitize.js]: I do not have access to any IE, but if you already tested this with IE, then it works for me

**Comment by IgorMinar** [comment]
you didn't update docs and used the wrong object to throw the error.

pull the commit from this branch: https://github.com/angular/angular.js/compare/master...IgorMinar:sanitize-fix?expand=1

review the docs for correctness..

**Comment by mhevery** [comment]
@IgorMinar merged your changes here, you can close your branch.

**Comment by lgalfaso** [comment]
The sanitization process looks fine, but this PR makes some unexpected html behavior changes that change how html is rendered. eg

```js
it('should not duplicate <br> tags', function() {
  expectHTML('<br />').not.toEqual('<br></br>');
  expectHTML('<br>').not.toEqual('<br></br>');
});
```

**Comment by IgorMinar** [comment]
This PR has been superseded by #12524

**Pull request closed without merging**